/*
 * Copyright (C) 2024 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

syntax = "proto2";
package android.aconfigd;
option optimize_for = LITE_RUNTIME;

// incoming request to aconfigd
message StorageRequestMessage {
  // new storage notification
  message NewStorageMessage {
    optional string container = 1;
    optional string package_map = 2;
    optional string flag_map = 3;
    optional string flag_value = 4;
  }

  // request persistent flag value override
  message FlagOverrideMessage {
    optional string package_name = 1;
    optional string flag_name = 2;
    optional string flag_value = 3;
    optional bool is_local = 4;
  }

  // request to remove local flag override
  message RemoveLocalOverrideMessage {
    optional bool remove_all = 1;
    optional string package_name = 2;
    optional string flag_name = 3;
  }

  // query persistent flag value and info
  message FlagQueryMessage {
    optional string package_name = 1;
    optional string flag_name = 2;
  }

  // reset all storage
  message ResetStorageMessage {}

  oneof msg {
    NewStorageMessage new_storage_message = 1;
    FlagOverrideMessage flag_override_message = 2;
    FlagQueryMessage flag_query_message = 3;
    RemoveLocalOverrideMessage remove_local_override_message = 4;
    ResetStorageMessage reset_storage_message = 5;
  };
}

message StorageRequestMessages {
  repeated StorageRequestMessage msgs = 1;
}

// aconfigd return to client
message StorageReturnMessage {
  message NewStorageReturnMessage {
    optional bool storage_updated = 1;
  }

  message FlagOverrideReturnMessage {}

  message FlagQueryReturnMessage {
    optional string server_flag_value = 1;
    optional string local_flag_value = 2;
    optional bool has_server_override = 3;
    optional bool is_readwrite = 4;
    optional bool has_local_override = 5;
  }

  message RemoveLocalOverrideReturnMessage {}

  oneof msg {
    NewStorageReturnMessage new_storage_message = 1;
    FlagOverrideReturnMessage flag_override_message = 2;
    FlagQueryReturnMessage flag_query_message = 3;
    RemoveLocalOverrideReturnMessage remove_local_override_message = 4;
    string error_message = 5;
  };
}

message StorageReturnMessages {
  repeated StorageReturnMessage msgs = 1;
}

message LocalFlagOverride {
  optional string package_name = 1;
  optional string flag_name = 2;
  optional string flag_value = 3;
}

message LocalFlagOverrides {
  repeated LocalFlagOverride overrides = 1;
}
